README.rst
.. image:: https://travis-ci.org/prisae/fftlog.svg?branch=master
   :target: https://travis-ci.org/prisae/fftlog
   :alt: Travis-CI
.. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.3830534.svg
   :target: https://doi.org/10.5281/zenodo.3830534
   :alt: Zenodo DOI


`fftlog` - A python wrapper for FFTLog
======================================

This is a simple `f2py`-wrapper for the logarithmic FFT code *FFTLog* as
presented in Appendix B of [Hami00]_ and published at
`casa.colorado.edu/~ajsh/FFTLog <http://casa.colorado.edu/~ajsh/FFTLog>`_.

A pure python version (`pyfftlog`) can be found on `github.com/prisae/pyfftlog
<https://github.com/prisae/pyfftlog>`_. Tests have shown that `fftlog` is a bit
faster than `pyfftlog`, but `pyfftlog` is easier to implement, as you only need
`NumPy` and `SciPy`, without the need to compile anything.

I hope that `FFTLog` will make it into `SciPy` in the future, which will make
this project redundant. (If you have the bandwidth and are willing to chip in
have a look at `SciPy PR #7310 <https://github.com/scipy/scipy/pull/7310>`_.)

Be aware that `pyfftlog` has not been tested extensively. It works fine for the
test from the original code, and my use case, which is `pyfftlog.fftl` with
`mu=0.5` (sine-transform), `q=0` (unbiased), `k=1`, `kropt=1`, and `tdir=1`
(forward). Please let me know if you encounter any issues.

- **Documentation**: https://pyfftlog.readthedocs.io
- **Source Code**: https://github.com/prisae/fftlog

**Note** that the documentation is for the pure python version `pyfftlog`, but
equally applies to `fftlog`.

Description of FFTLog from the FFTLog-Website
---------------------------------------------

FFTLog is a set of fortran subroutines that compute the fast Fourier or Hankel
(= Fourier-Bessel) transform of a periodic sequence of logarithmically spaced
points.

FFTLog can be regarded as a natural analogue to the standard Fast Fourier
Transform (FFT), in the sense that, just as the normal FFT gives the exact (to
machine precision) Fourier transform of a linearly spaced periodic sequence, so
also FFTLog gives the exact Fourier or Hankel transform, of arbitrary order m,
of a logarithmically spaced periodic sequence.

FFTLog shares with the normal FFT the problems of ringing (response to sudden
steps) and aliasing (periodic folding of frequencies), but under appropriate
circumstances FFTLog may approximate the results of a continuous Fourier or
Hankel transform.

The FFTLog algorithm was originally proposed by [Talm78]_.

*For the full documentation, see* `casa.colorado.edu/~ajsh/FFTLog
<http://casa.colorado.edu/~ajsh/FFTLog>`_.


Installation
------------

You can install fftlog either via **conda**:

.. code-block:: console

   conda install -c conda-forge fftlog

or via **pip**:

.. code-block:: console

   pip install fftlog


Creation
--------

The power of `f2py` did most of the work.

The *src*-directory contains the original fortran files as downloaded from
`casa.colorado.edu/~ajsh/FFTLog <http://casa.colorado.edu/~ajsh/FFTLog>`_.  The
only change I made was to recode the coding of *fftlog.f*, as `f2py` struggled
with a few characters in the description part:

.. code:: bash

   recode latin1..UTF-8 fftlog.f

Thereafter I used `f2py` to produce the `pyf`-instructions with the following
command, generating only hooks for the functions `fhti`, `fttl`, `fht`, and
`fhtq`:

.. code:: bash

   f2py src/* -m fftlog -h fftlog.pyf only: fhti fftl fht fhtq :

Lastly I amended the `pyf`-instructions, mainly with some `intent` and
`optional` statements as well as the corresponding default values.


Notes
'''''
1. `kropt=3` (interactive adjusting) is not possible with `fftlog`
2. `wsave`-dimension is set to `2*n+3*(n/2)+19`, the biggest of the four
   minimum sizes described in `fftlog.f`.


References
----------

.. [Hami00] Hamilton, A. J. S., 2000, Uncorrelated modes of the non-linear
    power spectrum: Monthly Notices of the Royal Astronomical Society, 312,
    pages 257-284; DOI: `10.1046/j.1365-8711.2000.03071.x
    <http://dx.doi.org/10.1046/j.1365-8711.2000.03071.x>`_; Website of FFTLog:
    `casa.colorado.edu/~ajsh/FFTLog <http://casa.colorado.edu/~ajsh/FFTLog>`_.

.. [Talm78] Talman, J. D., 1978, Numerical Fourier and Bessel transforms in
    logarithmic variables: Journal of Computational Physics, 29, pages 35-48;
    DOI: `10.1016/0021-9991(78)90107-9
    <http://dx.doi.org/10.1016/0021-9991(78)90107-9>`_.


License, Citation, and Credits
------------------------------

These additions to the original FFTLog-code are released to the public domain
under the `CC0 1.0 License
<http://creativecommons.org/publicdomain/zero/1.0>`_.

All releases have a Zenodo-DOI, which can be found on `10.5281/zenodo.3830364
<https://doi.org/10.5281/zenodo.3830534>`_.

Permission to distribute the original Fortran `FFTLog` code with this Python
`fftlog` package has been granted (email from Andrew Hamilton to Dieter
Werthmüller dated 28 September 2016).

Credits commented in the original code:

`FFTLog` uses the NCAR suite of FFT routines, and a modified version of the
complex Gamma function from the gamerf package at
`momonga.t.u-tokyo.ac.jp/~ooura/gamerf.html
<http://momonga.t.u-tokyo.ac.jp/~ooura/gamerf.html>`_.
The original gamerf copyright statement states::

   Copyright(C) 1996 Takuya OOURA (email: ooura@mmm.t.u-tokyo.ac.jp).
   You may use, copy, modify this code for any purpose and
   without fee. You may distribute this ORIGINAL package.

Permission to distribute the modified gamma function code with the FFTLog
package has been granted (email from Takuya Ooura to Andrew Hamilton dated 16
March 1999).

Be kind and give credits by citing Hamilton (2000).


CHANGELOG.rst
Changelog
#########

v0.2.1 : Fix packaging
----------------------

Version 0.2.0 had some problems with the way it was packaged on PyPi and
conda-forge. Nothing changed in the code itself.


v0.2.0 : First packaged release
-------------------------------

**2020-05-16**

First packaged release on PyPi and conda-forge. This includes:

- Re-structuring the repo.
- Link to proper documentation, https://pyfftlog.readthedocs.io
  (which is for pyfftlog), which contains as well the example.
- Add tests and CI on Travis, https://travis-ci.org/github/prisae/pyfftlog.
- Link to Zenodo, https://zenodo.org/record/3830366.
- PEP8 checking.
- Add the relevant badges to README.


v0.1.0 : Initial upload to GitHub
---------------------------------

**2016-12-05**

- Initially working version uploaded to GitHub.


tests/test_pyfftlog.py
import numpy as np
from numpy.testing import assert_allclose

from fftlog import fhti, fht


def test_fftlog():
    # This is the example provided in `fftlogtest.f` of the original code. It
    # is the same test that is shown as an example in the gallery.
    # See the example for more details.

    # Parameters.
    logrmin = -4
    logrmax = 4
    n = 64
    mu = 0
    q = 0
    kr = 1
    kropt = 1
    tdir = 1
    logrc = (logrmin + logrmax)/2
    nc = (n + 1)/2.0
    dlogr = (logrmax - logrmin)/n
    dlnr = dlogr*np.log(10.0)

    # Calculate input function: $r^{\mu+1}\exp\left(-\frac{r^2}{2}\right)$.
    r = 10**(logrc + (np.arange(1, n+1) - nc)*dlogr)
    ar = r**(mu + 1)*np.exp(-r**2/2.0)

    # Initialize FFTLog transform - note fhti resets `kr`
    kr, xsave, _ = fhti(n, mu, dlnr, q, kr, kropt)
    assert_allclose(kr, 0.953538967579)
    logkc = np.log10(kr) - logrc
    # rk = 10**(logrc - logkc)

    # Transform
    # ak = fftl(ar.copy(), xsave, rk, tdir)
    ak = fht(ar.copy(), xsave, tdir)

    # Calculate Output function: $k^{\mu+1}\exp\left(-\frac{k^2}{2}\right)$
    k = 10**(logkc + (np.arange(1, n+1) - nc)*dlogr)
    theo = k**(mu + 1)*np.exp(-k**2/2.0)

    # Check values
    assert_allclose(theo[theo > 1e-3], ak[theo > 1e-3], rtol=1e-8, atol=5e-5)


fftlogtest.py

# coding: utf-8

# # FFTLog
# 
# This notebook is a translation of `fftlogtest.f` from the Fortran package `FFTLog`, which was presented in Appendix B of Hamilton, 2000, and published at <http://casa.colorado.edu/~ajsh/FFTLog>. It serves as an example for the python package `fftlog` (which is a `f2py`-wrapper around `FFTLog`), in the same manner as the original file `fftlogtest.f` serves as an example for Fortran package `FFTLog`.
# 
# 
# ### Reference
# Hamilton, A. J. S., 2000, Uncorrelated modes of the non-linear power spectrum: Monthly Notices of the Royal Astronomical Society, 312, pages 257-284; DOI: <http://dx.doi.org/10.1046/j.1365-8711.2000.03071.x>.
# 
# ---

# ## This is fftlogtest.f
# 
# This is a simple test program to illustrate how `FFTLog` works.
# The test transform is:
# 
# $$
# \int^\infty_0 r^{\mu+1} \exp\left(-\frac{r^2}{2} \right)\
# J_\mu(k, r)\ k\ {\rm d}r = k^{\mu+1} \exp\left(-\frac{k^2}{2}
# \right) $$
# 
# 
# **Disclaimer:**  
# `FFTLog` does NOT claim to provide the most accurate possible
# solution of the continuous transform (which is the stated aim
# of some other codes).  Rather, `FFTLog` claims to solve the exact
# discrete transform of a logarithmically-spaced periodic sequence.
# If the periodic interval is wide enough, the resolution high
# enough, and the function well enough behaved outside the periodic
# interval, then `FFTLog` may yield a satisfactory approximation
# to the continuous transform.
# 
# Observe:
# 1.  How the result improves as the periodic interval is enlarged.
#     With the normal FFT, one is not used to ranges orders of
#     magnitude wide, but this is how `FFTLog` prefers it.
# 2.  How the result improves as the resolution is increased.
#     Because the function is rather smooth, modest resolution
#     actually works quite well here.
# 3.  That the central part of the transform is more reliable
#     than the outer parts.  Experience suggests that a good general
#     strategy is to double the periodic interval over which the
#     input function is defined, and then to discard the outer
#     half of the transform.
# 4.  That the best bias exponent seems to be $q = 0$.
# 5.  That for the critical index $\mu = -1$, the result seems to be
#     offset by a constant from the 'correct' answer.
# 6.  That the result grows progressively worse as mu decreases
#     below -1.
# 
# The analytic integral above fails for $\mu \le -1$, but `FFTLog`
# still returns answers.  Namely, `FFTLog` returns the analytic
# continuation of the discrete transform.  Because of ambiguity
# in the path of integration around poles, this analytic continuation
# is liable to differ, for $\mu \le -1$, by a constant from the 'correct'
# continuation given by the above equation.
# 
# `FFTLog` begins to have serious difficulties with aliasing as
# $\mu$ decreases below $-1$, because then $r^{\mu+1} \exp(-r^2/2)$ is
# far from resembling a periodic function.
# You might have thought that it would help to introduce a bias
# exponent $q = \mu$, or perhaps $q = \mu+1$, or more, to make the
# function $a(r) = A(r) r^{-q}$ input to `fhtq` more nearly periodic.
# In practice a nonzero $q$ makes things worse.
# 
# A symmetry argument lends support to the notion that the best
# exponent here should be $q = 0,$ as empirically appears to be true.
# The symmetry argument is that the function $r^{\mu+1} \exp(-r^2/2)$
# happens to be the same as its transform $k^{\mu+1} \exp(-k^2/2)$.
# If the best bias exponent were q in the forward transform, then
# the best exponent would be $-q$ that in the backward transform;
# but the two transforms happen to be the same in this case,
# suggesting $q = -q$, hence $q = 0$.
# 
# This example illustrates that you cannot always tell just by
# looking at a function what the best bias exponent $q$ should be.
# You also have to look at its transform.  The best exponent $q$ is,
# in a sense, the one that makes both the function and its transform
# look most nearly periodic.
# 
# ---

# ## Test-Integral:  $\int_0^\infty r^{\mu+1}\ \exp\left(-\frac{r^2}{2}\right)\ J_\mu(k,r)\ k\ {\rm d}r = k^{\mu+1} \exp\left(-\frac{k^2}{2}\right)$

# ### Import `fftlog` as well as `numpy` and `matplotlib`; some plot settings

# In[1]:

import fftlog
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

get_ipython().magic('matplotlib inline')
plt.style.use('ggplot')
mpl.rcParams.update({'font.size': 16})


# ### Define the parameters you wish to use
# 
# The presets are the *Reasonable choices of parameters* from `fftlogtest.f`.

# In[2]:

# Range of periodic interval
logrmin = -4
logrmax = 4

# Number of points (Max 4096)
n = 64

# Order mu of Bessel function
mu = 0

# Bias exponent: q = 0 is unbiased
q = 0

# Sensible approximate choice of k_c r_c
kr = 1

# Tell fhti to change kr to low-ringing value
# WARNING: kropt = 3 will fail, as interaction is not supported
kropt = 1

# Forward transform (changed from dir to tdir, as dir is a python fct)
tdir = 1


# ### Calculation related to the logarithmic spacing

# In[3]:

# Central point log10(r_c) of periodic interval
logrc = (logrmin + logrmax)/2

print('Central point of periodic interval at log10(r_c) = ', logrc)

# Central index (1/2 integral if n is even)
nc = (n + 1)/2.0

# Log-spacing of points
dlogr = (logrmax - logrmin)/n
dlnr = dlogr*np.log(10.0)


# ### Calculate input function: $r^{\mu+1}\exp\left(-\frac{r^2}{2}\right)$

# In[4]:

r = 10**(logrc + (np.arange(1, n+1) - nc)*dlogr)
ar = r**(mu + 1)*np.exp(-r**2/2.0)


# ### Initialize FFTLog transform - note fhti resets `kr`

# In[5]:

kr, wsave, ok = fftlog.fhti(n, mu, dlnr, q, kr, kropt)
print('fftlog.fhti: ok =', bool(ok), '; New kr = ', kr)


# ### Call `fftlog.fht` (or `fftlog.fhtl`)

# In[6]:

logkc = np.log10(kr) - logrc
print('Central point in k-space at log10(k_c) = ', logkc)

# rk = r_c/k_c
rk = 10**(logrc - logkc)

# Transform
#ak = fftlog.fftl(ar.copy(), wsave, rk, tdir)
ak = fftlog.fht(ar.copy(), wsave, tdir)


# ### Calculate Output function: $k^{\mu+1}\exp\left(-\frac{k^2}{2}\right)$

# In[7]:

k = 10**(logkc + (np.arange(1, n+1) - nc)*dlogr)
theo = k**(mu + 1)*np.exp(-k**2/2.0)


# ### Plot result

# In[8]:

plt.figure(figsize=(16,8))

# Transformed result
ax2 = plt.subplot(1, 2, 2)
plt.plot(k, theo, 'k', lw=2, label='Theoretical')
plt.plot(k, ak, 'r--', lw=2, label='FFTLog')
plt.xlabel('k')
plt.title(r'$k^{\mu+1} \exp(-k^2/2)$', fontsize=20)
plt.legend(loc='best')
plt.xscale('log')
plt.yscale('symlog', basey=10, linthreshy=1e-5)
ax2ylim = plt.ylim()

# Input
ax1 = plt.subplot(1, 2, 1)
plt.plot(r, ar, 'k', lw=2)
plt.xlabel('r')
plt.title(r'$r^{\mu+1}\ \exp(-r^2/2)$', fontsize=20)
plt.xscale('log')
plt.yscale('symlog', basey=10, linthreshy=1e-5)
plt.ylim(ax2ylim)

# Main title
plt.suptitle(r'$\int_0^\infty r^{\mu+1}\ \exp(-r^2/2)\ J_\mu(k,r)\ k\ {\rm d}r = k^{\mu+1} \exp(-k^2/2)$',
             fontsize=24, y=1.08)
plt.show()


# ### Print values

# In[9]:

print('           k                 a(k)       k^(mu+1) exp(-k^2/2)')
print('----------------------------------------------------------------')
for i in range(n):
    print("%18.6e %18.6e %18.6e"% (k[i], ak[i], theo[i]))

